### PR TITLE
refactor(menu): migrate Menu to Base UI

### PR DESCRIPTION
**Pull request by kylemcd**
## Stack context

- Part of the Telegraph Radix to Base UI migration stack.
- Parent PR: #842.
- Linear: [KNO-13670](https://linear.app/knock/issue/KNO-13670).

## What changed

- Migrates `@telegraph/menu` from Radix Menu to Base UI Menu while preserving the existing compound API.
- Reuses shared overlay compatibility helpers from `@telegraph/helpers` and keeps Menu-specific keyboard compatibility in `Menu.helpers.ts`.
- Preserves trigger/content refs, `tgphRef`, Radix-compatible popper CSS variables, preventable dismissal, focus return, submenu behavior, and item selection semantics.
- Preserves Combobox usage of Menu buttons, including `closeOnSelect={false}` multi-select behavior and accessible-label fallback when an option label is an empty string.
- Scopes menu content dismissal callbacks so root dismissal handlers do not block submenu Escape/close behavior.
- Preserves nested focusable trigger keyboard behavior without suppressing Base UI handlers unnecessarily.
- Updates Menu docs, tests, dependencies, changeset, and the Storybook script to avoid auto-opening a browser.

## Why

- Removes Menu's Radix runtime dependency while keeping published behavior and downstream Combobox usage stable.
- Keeps only the Telegraph-specific keyboard and dismissal edge cases local to Menu.
- Ensures stacked/nested menu interactions and Combobox menu-item selection keep matching the pre-migration behavior after moving to Base UI.

## Verification

- `yarn test packages/menu/src/Menu/Menu.test.tsx`
- `yarn test packages/combobox/src/Combobox/Combobox.test.tsx -t getOptionAccessibleLabel`
- `yarn workspace @telegraph/menu build`
- Latest stack-wide pass: `yarn format:check`, `yarn manypkg:check`, `yarn build:packages`, `yarn test`, `yarn lint`, `yarn build:storybook`, `git diff --check`

### CODE REVIEW

**Comment by kylemcd** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.com/github/pr/knocklabs/telegraph/843?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#851** <a href="https://app.graphite.com/github/pr/knocklabs/telegraph/851?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#850** <a href="https://app.graphite.com/github/pr/knocklabs/telegraph/850?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#849** <a href="https://app.graphite.com/github/pr/knocklabs/telegraph/849?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#848** <a href="https://app.graphite.com/github/pr/knocklabs/telegraph/848?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#847** <a href="https://app.graphite.com/github/pr/knocklabs/telegraph/847?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#846** <a href="https://app.graphite.com/github/pr/knocklabs/telegraph/846?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#845** <a href="https://app.graphite.com/github/pr/knocklabs/telegraph/845?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#844** <a href="https://app.graphite.com/github/pr/knocklabs/telegraph/844?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#843** <a href="https://app.graphite.com/github/pr/knocklabs/telegraph/843?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/knocklabs/telegraph/843?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#842** <a href="https://app.graphite.com/github/pr/knocklabs/telegraph/842?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#841** <a href="https://app.graphite.com/github/pr/knocklabs/telegraph/841?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#840** <a href="https://app.graphite.com/github/pr/knocklabs/telegraph/840?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#839** <a href="https://app.graphite.com/github/pr/knocklabs/telegraph/839?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#838** <a href="https://app.graphite.com/github/pr/knocklabs/telegraph/838?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#837** <a href="https://app.graphite.com/github/pr/knocklabs/telegraph/837?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: packages/combobox/src/Combobox/Combobox.tsx [hunk 1]
```diff
@@ -37,6 +53,22 @@ const FIRST_KEYS = ["ArrowDown", "PageUp", "Home"];
...
+    typeof option.label === "string" ||
+    typeof option.label === "number" ||
+    typeof option.label === "bigint"
+  ) {
+    return String(option.label);
+  }
+
+  return option.value;
+};
+
```
Comment (kylemcd) [inline packages/combobox/src/Combobox/Combobox.tsx]: this should be moved to the helpers file.

File: packages/menu/src/Menu/Menu.helpers.ts [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+export const composeRefs = <T>(...refs: Array<Ref<T> | undefined>) => {
+  if (!refs.some(Boolean)) {
+    return undefined;
+  }
+
+  return (node: T | null) => {
+    refs.forEach((ref) => setRef(ref, node));
+  };
+};
+
```
Comment (kylemcd) [inline packages/menu/src/Menu/Menu.helpers.ts]: these are used in another PR so we should def move them to helpers package. generally we should audit for duplicate helpers like these and consolidate them.

File: packages/menu/src/Menu/Menu.helpers.ts [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+export const getRootContentCallbacks = (
+  contentCallbacksRef: MenuContentCallbacksRef,
+) => {
+  return contentCallbacksRef.current[0]?.callbacks ?? {};
+};
+
```
Comment (kylemcd) [inline packages/menu/src/Menu/Menu.helpers.ts:77]: why do we need all of these keyboard helpers? is there a change between radix and base ui that is big enough to warrant all of this custom handling? is there a good reason not to just lean on base-ui here?

File: packages/menu/src/Menu/Menu.tsx [hunk 4]
```diff
@@ -1,322 +1,782 @@
...
+      ...props
+    },
+    forwardedRef,
+  ) => {
+    const triggerRef = composeRefs(
+      forwardedRef,
+      tgphRef as Ref<HTMLElement> | undefined,
+    );
+    const shouldSuppressBaseMouseDown = Boolean(onClick);
+    const handleClick = useCallback<
```
Comment (kylemcd) [inline packages/menu/src/Menu/Menu.tsx:190]: same with all of the handle functions, why aren't we just leaning on base-ui?

File: packages/menu/src/Menu/Menu.tsx [hunk 5]
```diff
@@ -1,322 +1,782 @@
...
-  onCloseAutoFocus,
+  sticky,
+  style,
   tgphRef,
   ...props
 }: ContentProps<T>) => {
+  const compatibilityContext = useContext(MenuCompatibilityContext);
+  const contentCallbacksIdRef = useRef(Symbol("menu-content-callbacks"));
+  const internalContentRef = useRef<HTMLDivElement>(null);
+  const contentRef = composeRefs(tgphRef, contentStackRef, internalContentRef);
```
Comment (kylemcd) [inline packages/menu/src/Menu/Menu.tsx]: tons more stuff in this file that feel like junk? can we audit what's going on with this component and figure out how we can cut out things so that we can lean on base-ui fully.

File: packages/combobox/src/Combobox/Combobox.tsx (hunk 1 above)
Comment (kylemcd) [reply packages/combobox/src/Combobox/Combobox.tsx]: Moved this helper-shaped Combobox logic into packages/combobox/src/Combobox/Combobox.helpers.ts in the Combobox layer, so the main component no longer owns those declarations.

File: packages/menu/src/Menu/Menu.helpers.ts (hunk 3 above)
Comment (kylemcd) [reply packages/menu/src/Menu/Menu.helpers.ts:77]: Audited this and trimmed it to compatibility behavior Base UI does not cover exactly: Telegraph/Radix-era focus movement from nested focusable children, prevented handler propagation, and legacy PageUp/Home/Arrow handling. Ordinary roving focus stays with Base UI; the remaining helper file has a why-comment and test coverage.

File: packages/menu/src/Menu/Menu.tsx (hunk 4 above)
Comment (kylemcd) [reply packages/menu/src/Menu/Menu.tsx:190]: Audited the handlers too. The remaining wrappers invoke consumer callbacks first, respect defaultPrevented via preventBaseUIHandler, and bridge Telegraph edge cases before Base UI handlers run. Shared dismissal helpers were consolidated into @telegraph/helpers.

**Review by meryldakin — APPROVED** [review]

**Comment by kylemcd** [comment]
Folded into #837 so the full Base UI migration can be reviewed and merged as a single PR.
